### PR TITLE
Fix behaviors that were staying in the list after a project switch

### PR DESCRIPTION
**Pull request by D8H**
* It happened if at least one extension tab was opened.

Extensions should not be reloaded on UI events when the project is closing.

### CODE REVIEW

File: newIDE/app/src/MainFrame/index.js [hunk 1]
```diff
@@ -351,6 +351,7 @@ const MainFrame = (props: Props) => {
   const [isProjectOpening, setIsProjectOpening] = React.useState<boolean>(
     false
   );
+  const [isProjectClosed, setIsProjectClosed] = React.useState<boolean>(false);
```
Comment (D8H) [inline newIDE/app/src/MainFrame/index.js:354]: I'm not sure if it's a good idea to have boolean for states.

Comment (ClementPasteau) [reply newIDE/app/src/MainFrame/index.js:354]: Having boolean for states is ok usually. 
Here I'm unsure about the meaning of this one:
- Are you able to do the same by just looking at the `project`? (as in `if (!project)` being the same as `isProjectClosed` ?
- If not, then the name should probably be different, like the boolean just above `isProjectClosing` ? Then you can set it to true at the beginning of the  `closeProject` function and set it to false at the end with a big `try` / `catch` / `finally` <- in the finally ?

Comment (D8H) [reply newIDE/app/src/MainFrame/index.js:354]: > * Are you able to do the same by just looking at the `state.currentProject`? (as in `if (!currentProject)` being the same as `isProjectClosed` ?
> * If not, then the name should probably be different, like the boolean just above `isProjectClosing` ? Then you can set it to true at the beginning of the  `closeProject` function and set it to false at the end with a big `try` / `catch` / `finally` <- in the finally ?

I wanted to do `isProjectClosing` at 1st but it seems complicated to know when a project can be considered closed:
- change the state to close tabs
- tabs detect a change of focus when they are closed and send an event
- the mainframe handle this event and choose to reload extension or not

What if `isProjectClosing` is set to false too early and events still refresh extensions when they shouldn't?
It seems risky.

Maybe `isProjectClosingOrClosed` to be more precised?

Comment (4ian) [reply newIDE/app/src/MainFrame/index.js:354]: I agree that we should aim for not having this state if we can, by looking at `currentProject`. If a project is not opened/closed, then `currentProject` should be null.

File: newIDE/app/src/MainFrame/index.js [hunk 2]
```diff
@@ -2902,6 +2906,9 @@ const MainFrame = (props: Props) => {
             }
             eventsFunctionsExtensionsError={eventsFunctionsExtensionsError}
             onReloadEventsFunctionsExtensions={() => {
+              if (isProjectClosed) {
+                return;
+              }
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js:2911]: This is equivalent to what `reloadProjectEventsFunctionsExtensions` is doing if the project passed is null.
We should prefer this because typing is stronger than "yet another boolean state".

File: newIDE/app/src/MainFrame/index.js [hunk 3]
```diff
@@ -3035,15 +3042,23 @@ const MainFrame = (props: Props) => {
                     onOpenPreferences: () => openPreferencesDialog(true),
                     onOpenAbout: () => openAboutDialog(true),
                     selectInAppTutorial: selectInAppTutorial,
-                    onLoadEventsFunctionsExtensions: () =>
-                      eventsFunctionsExtensionsState.loadProjectEventsFunctionsExtensions(
+                    onLoadEventsFunctionsExtensions: () => {
+                      if (isProjectClosed) {
+                        return Promise.resolve();
+                      }
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js:3048]: Same, if currentProject === null is equivalent to isProjectClosed, this is good already and we should strive to keep this.

File: newIDE/app/src/MainFrame/index.js (hunk 1 above)
Comment (D8H) [reply newIDE/app/src/MainFrame/index.js:354]: Splitting the setState in 2 is not a viable solution?

Comment (4ian) [reply newIDE/app/src/MainFrame/index.js:354]: I see three ideas:
- It's bad to start something when a component is unmounted. I remember this creating issues in the instances property. If you're unmounted, you probably don't want to call your parent, because maybe your parent will try to use a stale variable. In other words: **remove the componentWillUnmount**. Check how to still ensure extensions are reloaded if I do a change in an extension and just close the tab.
- We can consider it's the parent fault to try to do a job on something that was just set to null, but because they are staled they did not. It means the function where you added "isProjectClosed" should use `currentProjectRef`. Because this ref will always stay the same, and the variable inside is always up to date with the latest state.
  **I'm not sure if this will work, but try this**. **Log the value of `currentProjectRef.current`** and log when the `null` is set for currentProject and when it's destroyed. Finally, **log** when something is written by the extension loader.
  I think this might not work because if `currentProjectRef.current` is not `null`, then you still have the same issue: you're asking the extension loader to do a job on something that will be set to null and destroyed from memory very soon (at the next "tick" of the JS event loop).
- We can consider it's the responsibility of the project closing to WAIT FOR EXTENSIONS to be written. Use `await eventsFunctionsExtensionsState.ensureLoadFinished()`.

  In other words, this mean: "you asked the extension loader to do a job on a project, so you might as well wait for it to be done before trying to destroy the thing it is using! ;)"
  
Ideally, we would do all of these:
- avoid if we can to call things on unmounting. 
- even if we do or don't, check if the `currentProject` and `currentProjectRef.current` are different in the callbacks. I think they will be different. 
- even if we do or don't do this, let's still wait for `eventsFunctionsExtensionsState.ensureLoadFinished` because WHO KNOWS, maybe something has been triggering a extension loading, but somehow the hard drive took or network took 10 seconds to write a file, so a promise will resolve after the project is destroyed. So we must wait for this to finish.

File: newIDE/app/src/MainFrame/index.js [hunk 4]
```diff
@@ -2273,6 +2279,26 @@ const MainFrame = (props: Props) => {
     editorTab: EditorTab,
     newState: State = state
   ) => {
+    if (
```
Comment (4ian) [inline newIDE/app/src/MainFrame/index.js:2282]: That's indeed not super pretty/fragile. I don't like these new refs... 
Can we check why this code is not working as it should?

image

in newIDE/app/src/MainFrame/EditorContainers/EventsFunctionsExtensionEditorContainer.js.

Comment (D8H) [reply newIDE/app/src/MainFrame/index.js:2282]: I wanted to avoid the `componentWillUnmount` to call `onLoadEventsFunctionsExtensions`, I tried to:
- call `onLoadEventsFunctionsExtensions` when another tab is put upfront after an extension tab has been left
- Disable both `componentWillUnmount` (to avoid it on project closing) and `componentDidUpdate` (because it's already done)

The big issue is that the Home tab is handled differently because it's what will be shown when a project is closed. Still, users could open Home and launch a preview and I'm not sure the last edited extension would be up to date.

I'm not sure a clean solution can be done with only UI events. I think we should invest in a dirty flag system which will allow finer optimizations.

**Comment by D8H** [comment]
I reverted back to the initial solution to temporary fix the issue while we do something better.

File: newIDE/app/src/MainFrame/index.js (hunk 1 above)
Comment (4ian) [reply newIDE/app/src/MainFrame/index.js:354]: Let's name this state `isProjectClosedSoAvoidReloadingExtensions`.
It's a terrible, long name, and it's perfect because it should be a hint that if someone tries to use it, it's a terrible idea that was done to show that this should be removed as soon as we can.

**Review by 4ian — APPROVED** [review]
Ok with the state (and setter) named isProjectClosedSoAvoidReloadingExtensions. 

So no one is tempted to use this for something else.

**Review by 4ian — APPROVED** [review]
Thank you 👍
